libs/vkd3d/command.c

```diff
@@ -288,164 +288,101 @@ static HRESULT vkd3d_enqueue_timeline_semaphore(struct vkd3d_fence_worker *worke
-
-    ret = vkd3d_array_reserve((void **) &worker->fences, &worker->fences_size,
-                              count, sizeof(*worker->fences));
-
-    ret &= vkd3d_array_reserve((void **) &worker->vk_semaphores, &worker->vk_semaphores_size,
-                               count, sizeof(*worker->vk_semaphores));
-    ret &= vkd3d_array_reserve((void **) &worker->semaphore_wait_values, &worker->semaphore_wait_values_size,
-                               count, sizeof(*worker->semaphore_wait_values));
-
-    if (!ret)
-    {
-        ERR("Failed to reserve memory.\n");
-        return;
-    }
-
-    for (i = 0; i < worker->enqueued_fence_count; ++i)
-    {
-        struct vkd3d_enqueued_fence *current = &worker->enqueued_fences[i];
-
-        worker->vk_semaphores[worker->fence_count] = current->vk_semaphore;
-        worker->semaphore_wait_values[worker->fence_count] = current->waiting_fence.value;
-
-        worker->fences[worker->fence_count] = current->waiting_fence;
-        ++worker->fence_count;
-    }
-    assert(worker->fence_count == count);
-    worker->enqueued_fence_count = 0;
-}
-
-static void vkd3d_wait_for_gpu_timeline_semaphores(struct vkd3d_fence_worker *worker)
+static void vkd3d_wait_for_gpu_timeline_semaphore(struct vkd3d_fence_worker *worker, const struct vkd3d_enqueued_fence *fence)
 {
     struct d3d12_device *device = worker->device;
     const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
     VkSemaphoreWaitInfoKHR wait_info;
-    VkSemaphore vk_semaphore;
     uint64_t counter_value;
-    unsigned int i, j;
     HRESULT hr;
     int vr;
 
-    if (!worker->fence_count)
-        return;
-
     wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO_KHR;
     wait_info.pNext = NULL;
-    wait_info.flags = VK_SEMAPHORE_WAIT_ANY_BIT_KHR;
-    wait_info.pSemaphores = worker->vk_semaphores;
-    wait_info.semaphoreCount = worker->fence_count;
-    wait_info.pValues = worker->semaphore_wait_values;
+    wait_info.flags = 0;
+    wait_info.semaphoreCount = 1;
+    wait_info.pSemaphores = &fence->vk_semaphore;
+    wait_info.pValues = &fence->waiting_fence.value;
+
+    do
+    {
+        vr = VK_CALL(vkWaitSemaphoresKHR(device->vk_device, &wait_info, ~(uint64_t)0));
+    }
+    while (vr == VK_TIMEOUT);
 
-    vr = VK_CALL(vkWaitSemaphoresKHR(device->vk_device, &wait_info, ~(uint64_t)0));
-    if (vr == VK_TIMEOUT)
-        return;
     if (vr != VK_SUCCESS)
     {
-        ERR("Failed to wait for Vulkan timeline semaphores, vr %d.\n", vr);
+        ERR("Failed to wait for Vulkan timeline semaphore, vr %d.\n", vr);
         return;
     }
 
     /* This is a good time to kick the debug ring thread into action. */
     if (device->debug_ring.active)
         pthread_cond_signal(&device->debug_ring.ring_cond);
 
-    for (i = 0, j = 0; i < worker->fence_count; ++i)
+    if ((vr = VK_CALL(vkGetSemaphoreCounterValueKHR(device->vk_device, fence->vk_semaphore, &counter_value))) < 0)
```

Comment:
Is it even meaningful to get the counter value? We should just signal d3d12_fence_signal with the physical value we waited on.

Reply:
Yeah, will fix. I guess the old code did it to check which of the fences should be signalled?